### PR TITLE
fix missing connect timeout and make tests safer

### PR DESCRIPTION
**Pull request by artemredkin**
Fix missing connect timeout and make some tests faster and safer.

Motivation:
1. During pool implementation/refactoring we lost connection timeout, this PR fixes that
2. `testConnectionFailError` is slow, 10 seconds, due to not having a timeout set
3. `testConnectionFailError` and `testTLSVersionError` use dangerous `do-try-XCFail-catch` pattern

Modifications:
1. Set connection timeout if specified
2. Set connect timeout to 150 ms to make `testTLSVersionError` faster
3. Use `XCTAssertThrowsError` in `testConnectionFailError` and `testTLSVersionError`
4. Add general connection timeout test, since `testConnectionFailError` is NIOTS-specific.

Result:
Fixes an issue where connection timeout configuration was not respected, hopefully makes tests more robust

### CODE REVIEW

File: Tests/AsyncHTTPClientTests/HTTPClientNIOTSTests.swift [hunk 1]
```diff
@@ -74,19 +74,24 @@ class HTTPClientNIOTSTests: XCTestCase {
...
 
-        do {
-            _ = try httpClient.get(url: "https://localhost:\(port)/get").wait()
-            XCTFail("This should have failed")
-        } catch ChannelError.connectTimeout {
-        } catch {
-            XCTFail("Error should have been ChannelError.connectTimeout not \(type(of: error))")
+        XCTAssertThrowsError(try httpClient.get(url: "https://localhost:\(port)/get").wait()) { error in
+            switch error {
+            case ChannelError.connectTimeout(let timeout):
```
Comment (artemredkin) [inline Tests/AsyncHTTPClientTests/HTTPClientNIOTSTests.swift:90]: @weissi do you know if there is a safer pattern for this case, where enum has an associated value?

File: Tests/AsyncHTTPClientTests/HTTPClientTests.swift [hunk 2]
```diff
@@ -422,6 +422,25 @@ class HTTPClientTests: XCTestCase {
...
+    func testConnectTimeout() throws {
+        let httpClient = HTTPClient(eventLoopGroupProvider: .shared(self.clientGroup),
+                                    configuration: .init(timeout: .init(connect: .milliseconds(100), read: .milliseconds(150))))
+
+        defer {
+            XCTAssertNoThrow(try httpClient.syncShutdown())
+        }
+
+        // This must throw as 198.51.100.254 is reserved for documentation only
+        XCTAssertThrowsError(try httpClient.get(url: "http://198.51.100.254:65535/get").wait()) { error in
```
Comment (artemredkin) [inline Tests/AsyncHTTPClientTests/HTTPClientTests.swift:434]: @weissi I looked this up here:
https://github.com/apple/swift-nio/blob/6046aaa6291fb075c83b5c22bad41b9780b2d977/Tests/NIOTests/ChannelTests.swift#L1060
is this ok? or is there a better way to get a connect timeout?

**Comment by artemredkin** [comment]
failed due to #265

**Comment by artemredkin** [comment]
@swift-server-bot test this please

File: Tests/AsyncHTTPClientTests/HTTPClientNIOTSTests.swift (hunk 1 above)
Comment (Lukasa) [reply Tests/AsyncHTTPClientTests/HTTPClientNIOTSTests.swift:90]: What don't you like about this pattern?

File: Tests/AsyncHTTPClientTests/HTTPClientTests.swift (hunk 2 above)
Comment (Lukasa) [reply Tests/AsyncHTTPClientTests/HTTPClientTests.swift:434]: This should be just fine. Misconfigured networks can make this test fail, but without a complete user space IP and TCP implementation we don't have a better option.

File: Tests/AsyncHTTPClientTests/HTTPClientNIOTSTests.swift (hunk 1 above)
Comment (artemredkin) [reply Tests/AsyncHTTPClientTests/HTTPClientNIOTSTests.swift:90]: This is potentially dangerous, if you miss the `XCTFail` call in the default clause, you might miss when the test breaks. This is why I'm replacing `do-try-XCTFail-catch` pattern with `XCTAssertThrowsError`

Comment (Lukasa) [reply Tests/AsyncHTTPClientTests/HTTPClientNIOTSTests.swift:90]: I see what you mean. No, there isn't really a better pattern here sadly.

Comment (weissi) [reply Tests/AsyncHTTPClientTests/HTTPClientNIOTSTests.swift:90]: I usually do this

```
XCTAssertEqul(.connectTimeout(.milliSeconds(150)), error as? ChannelError)
```

as it's much safer

**Pull request merged**